flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServicesTest.java

```diff
@@ -0,0 +1,161 @@
+
+/**
+ * Tests auxiliary shared services created by {@link JobManagerSharedServices} and used by the
+ * {@link JobMaster}.
+ */
+public class JobManagerSharedServicesTest extends TestLogger {
+
+    private static final int NUM_FUTURE_THREADS = 8;
+    private static final int NUM_IO_THREADS = 4;
+    private static final int CPU_CORES = Hardware.getNumberCPUCores();
+
+    @Test
+    public void testFutureExecutorNoConfiguration() throws Exception {
+        Configuration config = new Configuration();
+
+        final JobManagerSharedServices jobManagerSharedServices =
+                JobManagerSharedServices.fromConfiguration(
+                        config,
+                        new BlobServer(config, new VoidBlobStore()),
+                        new TestingFatalErrorHandler());
+
+        CountDownLatch releasedLatch = new CountDownLatch(CPU_CORES);
+        CountDownLatch unreleasedLatch = new CountDownLatch(CPU_CORES + 1);
+        ScheduledExecutorService futureExecutor = jobManagerSharedServices.getFutureExecutor();
+        Runnable countsDown =
+                () -> {
+                    releasedLatch.countDown();
+                    unreleasedLatch.countDown();
+                };
+        for (int i = 0; i < CPU_CORES; i++) {
+            futureExecutor.execute(countsDown);
+        }
+        // Give the executor enough time to process all the Runnables
+        releasedLatch.await(1L, TimeUnit.SECONDS);
+        assertEquals(0, releasedLatch.getCount());
+        assertEquals(1, unreleasedLatch.getCount());
+        jobManagerSharedServices.shutdown();
+    }
+
+    @Test
+    public void testFutureExecutorConfiguration() throws Exception {
+        Configuration config = new Configuration();
+        config.setInteger(JobManagerOptions.JOB_MANAGER_FUTURE_THREADS, NUM_FUTURE_THREADS);
+
+        final JobManagerSharedServices jobManagerSharedServices =
+                JobManagerSharedServices.fromConfiguration(
+                        config,
+                        new BlobServer(config, new VoidBlobStore()),
+                        new TestingFatalErrorHandler());
+
+        CountDownLatch releasedLatch = new CountDownLatch(NUM_FUTURE_THREADS);
+        CountDownLatch unreleasedLatch = new CountDownLatch(NUM_FUTURE_THREADS + 1);
+        ScheduledExecutorService futureExecutor = jobManagerSharedServices.getFutureExecutor();
+        Runnable countsDown =
+                () -> {
+                    releasedLatch.countDown();
+                    unreleasedLatch.countDown();
+                };
+        for (int i = 0; i < NUM_FUTURE_THREADS; i++) {
+            futureExecutor.execute(countsDown);
+        }
+        // Give the executor enough time to process all the Runnables
+        releasedLatch.await(1L, TimeUnit.SECONDS);
+        assertEquals(0, releasedLatch.getCount());
+        assertEquals(1, unreleasedLatch.getCount());
+        jobManagerSharedServices.shutdown();
```

Comment:
We probably can deduplicate this part and the part in the `testFutureExecutorNoConfiguration`.

Reply:
Deduplicated!